### PR TITLE
Cleanup Concurrent RepositoryData Loading

### LINKED ISSUE #48122: [CI] SLMSnapshotBlockingIntegTests.testBasicPartialRetention failed
`SLMSnapshotBlockingIntegTests.testBasicPartialRetention` failed in https://gradle-enterprise.elastic.co/s/shm65wzp2dh54

The error is:

```
org.elasticsearch.xpack.slm.SLMSnapshotBlockingIntegTests > testBasicPartialRetention FAILED
org.elasticsearch.snapshots.SnapshotMissingException: [my-repo:snap-hfs4h9aptqkjnchyhkltla] is missing
```

This did not reproduce locally using:

```
./gradlew ':x-pack:plugin:ilm:test' --tests "org.elasticsearch.xpack.slm.SLMSnapshotBlockingIntegTests.testBasicPartialRetention" \
  -Dtests.seed=EAF550063D3EF24D \
  -Dtests.security.manager=true \
  -Dtests.locale=nn \
  -Dtests.timezone=America/Kentucky/Louisville \
  -Dcompiler.java=12 \
  -Druntime.java=11
```

### PR DESCRIPTION
**Pull request by original-brownbear**
The loading of `RepositoryData` is not an atomic operation.
It uses a list + get combination of calls.
This lead to accidentally returning an empty repository data
for generations >=0 which can never not exist unless the repository
is corrupted.
In the test #48122 (and other SLM tests) there was a low chance of
running into this concurrent modification scenario and the repository
actually moving two index generations between listing out the
index-N and loading the latest version of it. Since we only keep
two index-N around at a time this lead to unexpectedly absent
snapshots in status APIs.
Fixing the behavior to be more resilient is non-trivial but in the works.
For now I think we should simply throw in this scenario. This will also
help prevent corruption in the unlikely event but possible of running into this
issue in a snapshot create or delete operation on master failover on a
repository like S3 which doesn't have the "no overwrites" protection on
writing a new index-N.

I would suggest back porting this all the way since it can theoretically have 
repository corrupting effects on S3.

Fixes #48122

### CODE REVIEW

**Comment by original-brownbear** [comment]
I didn't yet get around to a full test run here, so if this doesn't yet go green please hold off reviewing till I can fix the tests I may have missed :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -812,9 +812,6 @@ public void endVerification(String seed) {
     public RepositoryData getRepositoryData() {
         try {
             return getRepositoryData(latestIndexBlobId());
-        } catch (NoSuchFileException ex) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: There is no good reason to catch here or below. We break out on generation `-1` and return empty data so whenever we fail to find data for gen. `>= 0` it's a problem and we mustn't ignore it.

**Comment by original-brownbear** [comment]
Jup, this is not entirely unsurprisingly failing some REST tests:

```
Caused by: org.elasticsearch.ElasticsearchException: Elasticsearch exception [type=no_such_file_exception, reason=/dev/shm/elastic+elasticsearch+pull-request-1/client/rest-high-level/build/testclusters/integTest-0/repo/index-16]
	at org.elasticsearch.ElasticsearchException.innerFromXContent(ElasticsearchException.java:496)
	at org.elasticsearch.ElasticsearchException.fromXContent(ElasticsearchException.java:407)
	at org.elasticsearch.ElasticsearchException.innerFromXContent(ElasticsearchException.java:437)
	at org.elasticsearch.ElasticsearchException.failureFromXContent(ElasticsearchException.java:603)
	at org.elasticsearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:169)
	... 45 more
```

-> WIP

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -925,9 +922,6 @@ private RepositoryData getRepositoryData(long indexGen) {
                 repositoryData = RepositoryData.snapshotsFromXContent(parser, indexGen);
             }
             return repositoryData;
-        } catch (NoSuchFileException ex) {
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: nit: while we are at it, there are unneeded `Long.toString()` and local variable `repositoryData` usages

**Review by tlrx — APPROVED** [review]
LGTM, thanks Armin

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I can't find it right now, but others disagreed with removing the `Long.toString` in another PR :( But yea let's clean up the redundant locals :)

**Comment by original-brownbear** [comment]
Thanks Tanguy!

**Pull request merged**